<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Use R to Create and Execute Reproducible CWL Workflows for Genomic Research • RcwlWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Use R to Create and Execute Reproducible CWL Workflows for Genomic Research">
<meta property="og:description" content="RcwlWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">RcwlWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ReUseData_Rcwl_Workshop.html">Use R to Create and Execute Reproducible CWL Workflows for Genomic Research</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rworkflow/RcwlWorkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Use R to Create and Execute Reproducible CWL
Workflows for Genomic Research</h1>
                        <h4 data-toc-skip class="author">Qian Liu[<a href="mailto:qian.liu@roswellpark.org" class="email">qian.liu@roswellpark.org</a>]</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/seandavi/BuildABiocWorkshop/blob/HEAD/vignettes/ReUseData_Rcwl_Workshop.Rmd" class="external-link"><code>vignettes/ReUseData_Rcwl_Workshop.Rmd</code></a></small>
      <div class="hidden name"><code>ReUseData_Rcwl_Workshop.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="use-r-to-create-and-execute-reproducible-cwl-workflows-for-genomic-research">Use R to Create and Execute Reproducible CWL Workflows for Genomic
Research<a class="anchor" aria-label="anchor" href="#use-r-to-create-and-execute-reproducible-cwl-workflows-for-genomic-research"></a>
</h2>
<p>Authors: Qian Liu <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, Another Author<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>. <br> Last modified:
July 27, 2023.</p>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<div class="section level4">
<h4 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h4>
<ul>
<li>Basic familiarity with DNA-seq data variant calling</li>
<li>Interest of using workflow language</li>
</ul>
</div>
<div class="section level4">
<h4 id="workshop-participation">Workshop Participation<a class="anchor" aria-label="anchor" href="#workshop-participation"></a>
</h4>
<p>The workshop format is a 45 minute session consisting of hands-on
demos, exercises and Q&amp;A.</p>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used">R / Bioconductor packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h4>
<ul>
<li>ReUseData</li>
<li>RcwlPipelines</li>
<li>Rcwl</li>
</ul>
</div>
<div class="section level4">
<h4 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h4>
<p>In this workshop, we will demonstrate how to use
<code>RcwlPipelines</code> and <code>ReUseData</code> to implement
variant calling workflows within R using the <code>Mutect2</code> from
GATK. This whole workflow is based on R programming language and can be
deployed in local computer, HPC and cloud computing platforms, using
<code>docker</code>, <code>singularity</code> or <code>udocker</code>.
[FIXME]</p>
<p>This workshop will be instructor-led live demo with executable code
with real DNA-seq data.</p>
<p>[DELETE] Along with the topic of your workshop, include how students
can expect to spend their time. For the description may also include
information about what type of workshop it is (e.g. instructor-led live
demo, lab, lecture + lab, etc.). Instructors are strongly recommended to
provide completely worked examples for lab sessions, and a set of
stand-alone notes that can be read and understood outside of the
workshop.</p>
</div>
<div class="section level4">
<h4 id="pre-requisites-1">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites-1"></a>
</h4>
<p>List any workshop prerequisites, for example:</p>
<ul>
<li>Basic knowledge of R syntax</li>
<li>Familiarity with the GenomicRanges class</li>
<li>Familiarity with xyz vignette (provide link)</li>
</ul>
<p>List relevant background reading for the workshop, including any
theoretical background you expect students to have.</p>
<ul>
<li>List any textbooks, papers, or other reading that students should be
familiar with. Include direct links where possible.</li>
</ul>
</div>
<div class="section level4">
<h4 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h4>
<p>Describe how students will be expected to participate in the
workshop.</p>
</div>
<div class="section level4">
<h4 id="r-bioconductor-packages-used-1">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used-1"></a>
</h4>
<pre><code>Rcwl
RcwlPipelines
ReUseData [FIXME: more?]</code></pre>
<p>List any <em>R</em> / <em>Bioconductor</em> packages that will be
explicitly covered.</p>
</div>
<div class="section level4">
<h4 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h4>
<p>An example for a 45-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Packages</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Package Development</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Contributing to Bioconductor</td>
<td>5m</td>
</tr>
<tr class="even">
<td>Best Practices</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h4>
<p>List “big picture” student-centered workshop goals and learning
objectives. Learning goals and objectives are related, but not the same
thing. These goals and objectives will help some people to decide
whether to attend the conference for training purposes, so please make
these as precise and accurate as possible.</p>
<p><em>Learning goals</em> are high-level descriptions of what
participants will learn and be able to do after the workshop is over.
<em>Learning objectives</em>, on the other hand, describe in very
specific and measurable terms specific skills or knowledge attained. The
<a href="#bloom">Bloom’s Taxonomy</a> may be a useful framework for
defining and describing your goals and objectives, although there are
others.</p>
</div>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<p>Some examples:</p>
<ul>
<li>describe how to…</li>
<li>identify methods for…</li>
<li>understand the difference between…</li>
</ul>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>analyze xyz data to produce…</li>
<li>create xyz plots</li>
<li>evaluate xyz data for artifacts</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="workshop-somatic-variant-calling">Workshop: Somatic variant calling<a class="anchor" aria-label="anchor" href="#workshop-somatic-variant-calling"></a>
</h3>
</div>
<div class="section level3">
<h3 id="workshop-downstream-variant-annotationfiltering-in-r">Workshop: downstream: variant annotation/filtering in R<a class="anchor" aria-label="anchor" href="#workshop-downstream-variant-annotationfiltering-in-r"></a>
</h3>
<p>For the somatic variant calling, we will need to prepare the
following:</p>
<ul>
<li>Experiment data
<ul>
<li>In the format of <code>.bam</code>, <code>.bam.bai</code> files</li>
</ul>
</li>
<li>ReUsable Genomic data
<ul>
<li>reference sequence file (<code>b37</code> or <code>hg38</code>)</li>
<li>Panel of Normals (PON) <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890631-Panel-of-Normals-PON-" class="external-link">ref</a>
</li>
</ul>
</li>
<li>Software tool:
<ul>
<li>Here we use <code>Mutect2</code>to Call somatic SNVs and indels via
local assembly of haplotypes. <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2" class="external-link">ref</a>
</li>
</ul>
</li>
</ul>
<p>There will be three ways to do the analysis task.</p>
<ol style="list-style-type: decimal">
<li><p>Command-line interface. We can install the software and write
shell scripts.</p></li>
<li>
<p>Workflow language. We can use workflow language (e.g., CWL, WDL,
nextflow, snakemake) to streamline the workflow of data analysis, using
containerized tools (no software installation, version tracking,
etc.)</p>
<ul>
<li>More efficient (especially if multiple tools are included)</li>
<li>REPRODUCIBLE!</li>
<li>Need to install a workflow runner (e.g., cwltool,
arvados-cwl-runner) [FIXME: install cwl?]</li>
<li>Learning curve.</li>
</ul>
</li>
<li>
<p><code>Rcwl/RcwlPipelines</code>. Write/use CWL-based data
analysis tools/workflows within <em>R</em> programming language</p>
<ul>
<li>All benefits of workflow language</li>
<li>Hands-on experience</li>
<li>Easy connection with downstream analysis tools in
<em>R/Bioconductor</em>.</li>
</ul>
<p><code>ReUseData</code>: Reproducible and reusable genomic data
management.</p>
<ul>
<li>Public genomic data resources can be easily managed and reused in
different projects.</li>
</ul>
</li>
</ol>
<p>In this workshop, we will demonstrate three <em>Bioconductor</em>
packages: <code>Rcwl</code> as an R interface for <code>CWL</code>;
<code>RcwlPipelines</code> with &gt;200 pre-built bioinformatics tools
and best practice pipelines in <em>R</em>, that are easily usable and
highly customizable; and <code>ReUseData</code> for the management of
reusable genomic data.</p>
<p>With these tools, we should be able to conduct reproducible data
analysis using commonly used bioinformatics tools (including
command-line based tools and <em>R/Bioconductor</em> packages) and
validated, best practice workflows (based on workflow languages such as
CWL) within a unified <em>R</em> programming environment.</p>
<div class="section level4">
<h4 id="prepare-experiment-data">Prepare experiment data<a class="anchor" aria-label="anchor" href="#prepare-experiment-data"></a>
</h4>
<p>Here we have prepared a small DNA-seq dataset for demo purposes. We
can download them from the GitHub repository and save to a specific
folder.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rworkflow/ReUseData" class="external-link">ReUseData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RcwlPipelines</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: Rcwl</span></span>
<span><span class="co">#&gt; Loading required package: yaml</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: BiocFileCache</span></span>
<span><span class="co">#&gt; Loading required package: dbplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'RcwlPipelines'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:graphics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     title</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Rcwl</span><span class="op">)</span></span>
<span><span class="va">workdir</span> <span class="op">&lt;-</span> <span class="st">"/tmp/Bioc2023_rcwl"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">workdir</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in dir.create(workdir): '/tmp/Bioc2023_rcwl' already exists</span></span></code></pre></div>
<p>To better manage the experiment data, we can write a data recipe and
add some meta information for them to be easily tracked.</p>
<p>The <code>recipeMake</code> function will wrap the command line
script for data downloading/processings into an executable data recipe
in R. We will specify the inputs and outputs of the data recipe and use
<code>outputGlob</code> to specify the output pattern (for internal
check).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">script</span> <span class="op">&lt;-</span> <span class="st">'</span></span>
<span><span class="st">wget https://github.com/hubentu/somatic-snv-test-data/raw/master/tumor.bam</span></span>
<span><span class="st">wget https://github.com/hubentu/somatic-snv-test-data/raw/master/tumor.bam.bai</span></span>
<span><span class="st">wget https://github.com/hubentu/somatic-snv-test-data/raw/master/normal.bam</span></span>
<span><span class="st">wget https://github.com/hubentu/somatic-snv-test-data/raw/master/normal.bam.bai</span></span>
<span><span class="st">'</span></span>
<span></span>
<span><span class="va">rcp_expdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/recipeMake.html" class="external-link">recipeMake</a></span><span class="op">(</span>shscript <span class="op">=</span> <span class="va">script</span>,</span>
<span>                          outputID <span class="op">=</span> <span class="st">"bams"</span>,</span>
<span>                          outputGlob <span class="op">=</span> <span class="st">"*.bam*"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## rcp_expdata &lt;- addMeta(</span></span>
<span><span class="co">##     cwl = rcp_expdata,</span></span>
<span><span class="co">##     label = "somatic snv test data",</span></span>
<span><span class="co">##     doc = "test data for somatic variant calling",</span></span>
<span><span class="co">##     outputLabels = c("bams"),</span></span>
<span><span class="co">##     outputDocs = c("The data is from NA12878 (tumor) and NA12892 (normal) and represents a tiny chunk of chromsome 21. It was created from 1000 Genomes Project."),</span></span>
<span><span class="co">##     extensions = list(</span></span>
<span><span class="co">##         author = "genome/somatic-snv-test-data",</span></span>
<span><span class="co">##         url = "https://github.com/genome/somatic-snv-test-data",</span></span>
<span><span class="co">##         date = Sys.Date()))</span></span></code></pre></div>
<p>Then we need to assign values to the input parameters if they exist
(data inputs for the shell script) and get ready for the evaluation. The
<code>getData</code> function evaluates the data recipe and generate the
desired data output in a user-specified <code>outdir</code>. Arbitrary
<code>notes</code> can be added for the dataset for easy query
later.</p>
<p>Internally, the data recipe is submitted as a cwl task evaluting the
shell script for data downloading/processing.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/getData.html" class="external-link">getData</a></span><span class="op">(</span><span class="va">rcp_expdata</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"data"</span><span class="op">)</span>,</span>
<span>        notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"somatic snv"</span>, <span class="st">"tumor normal bams"</span><span class="op">)</span>,</span>
<span>        showLog<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; List of length 4</span></span>
<span><span class="co">#&gt; names(4): command output logs yml</span></span></code></pre></div>
<p>We can see the files are successfully downloaded.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"data"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"*.bam*"</span><span class="op">)</span>  <span class="co">## desired experiment data</span></span>
<span><span class="co">#&gt; [1] "normal.bam"     "normal.bam.bai" "tumor.bam"      "tumor.bam.bai"</span></span></code></pre></div>
<p>Accompanying with the desired datasets, we can also see some files
with extensions of <code>md5</code>, <code>cwl</code>, <code>yml</code>
and <code>sh</code>. These files are automatically generated by the
function <code>getData</code>, where the the meta information for data
recipe are recorded.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"data"</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="st">"rcp_expdata"</span><span class="op">)</span>  <span class="co">## meta files</span></span>
<span><span class="co">#&gt; [1] "rcp_expdata_.cwl" "rcp_expdata_.md5" "rcp_expdata_.sh"  "rcp_expdata_.yml"</span></span></code></pre></div>
<p>By default, the meta file names for these meta files are paste of
recipe name, and input parameter values separated by <code>_</code>,
which can be changed by the <code>prefix</code> argument if needed.</p>
<ul>
<li>
<code>[recipeName_params].cwl</code>: Rcwl object defined in
<code>Rcwl</code> for tool or data recipe.</li>
<li>
<code>[recipeName_params].yml</code>: File containing values for
input parameters for both tool and data recipes through recipe
evaluation functions (e.g., <code>runCWL</code>, <code>getData</code>).
For data recipes, it includes additional meta information for the output
files, notes, and date, etc for data tracking purposes, which are added
by <code>getData</code> function.</li>
<li>
<code>[recipeName_params].sh</code>: The command lines in a shell
script.</li>
<li>
<code>[recipeName_params].md5</code>: unique identifier for each
dataset.</li>
</ul>
<div class="section level5">
<h5 id="data-use">Data use<a class="anchor" aria-label="anchor" href="#data-use"></a>
</h5>
<p>Then we use <code>dataUpdate</code> to cache the data in specified
data directory <code>dir</code>. Then the data can be easily tracked and
queried using <code>dataSearch</code> function. Meta information about
the data, such as file paths, can be easily retrieved for later use.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataUpdate.html" class="external-link">dataUpdate</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="va">workdir</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Updating data record...</span></span>
<span><span class="co">#&gt; normal.bam added</span></span>
<span><span class="co">#&gt; normal.bam.bai added</span></span>
<span><span class="co">#&gt; tumor.bam added</span></span>
<span><span class="co">#&gt; tumor.bam.bai added</span></span>
<span><span class="co">#&gt; Mutect2-exome-panel.vcf added</span></span>
<span><span class="co">#&gt; Mutect2-exome-panel.vcf.idx added</span></span>
<span><span class="co">#&gt; dataHub with 6 records</span></span>
<span><span class="co">#&gt; cache path:  /github/home/.cache/R/ReUseData </span></span>
<span><span class="co">#&gt; # dataUpdate() to update the local data cache</span></span>
<span><span class="co">#&gt; # dataSearch() to query a specific dataset</span></span>
<span><span class="co">#&gt; # Additional information can be retrieved using: </span></span>
<span><span class="co">#&gt; # dataNames(), dataParams(), dataNotes(), dataPaths(), dataTag() or mcols()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           name                       </span></span>
<span><span class="co">#&gt;   BFC11 | normal.bam                 </span></span>
<span><span class="co">#&gt;   BFC12 | normal.bam.bai             </span></span>
<span><span class="co">#&gt;   BFC13 | tumor.bam                  </span></span>
<span><span class="co">#&gt;   BFC14 | tumor.bam.bai              </span></span>
<span><span class="co">#&gt;   BFC15 | Mutect2-exome-panel.vcf    </span></span>
<span><span class="co">#&gt;   BFC16 | Mutect2-exome-panel.vcf.idx</span></span>
<span><span class="co">#&gt;         Path                                                    </span></span>
<span><span class="co">#&gt;   BFC11 /tmp/Bioc2023_rcwl/data/normal.bam                      </span></span>
<span><span class="co">#&gt;   BFC12 /tmp/Bioc2023_rcwl/data/normal.bam.bai                  </span></span>
<span><span class="co">#&gt;   BFC13 /tmp/Bioc2023_rcwl/data/tumor.bam                       </span></span>
<span><span class="co">#&gt;   BFC14 /tmp/Bioc2023_rcwl/data/tumor.bam.bai                   </span></span>
<span><span class="co">#&gt;   BFC15 /tmp/Bioc2023_rcwl/shareData/Mutect2-exome-panel.vcf    </span></span>
<span><span class="co">#&gt;   BFC16 /tmp/Bioc2023_rcwl/shareData/Mutect2-exome-panel.vcf.idx</span></span>
<span><span class="va">ds1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataSearch.html" class="external-link">dataSearch</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bam"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataPaths</a></span><span class="op">(</span><span class="va">ds1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/tmp/Bioc2023_rcwl/data/normal.bam"    </span></span>
<span><span class="co">#&gt; [2] "/tmp/Bioc2023_rcwl/data/normal.bam.bai"</span></span>
<span><span class="co">#&gt; [3] "/tmp/Bioc2023_rcwl/data/tumor.bam"     </span></span>
<span><span class="co">#&gt; [4] "/tmp/Bioc2023_rcwl/data/tumor.bam.bai"</span></span>
<span><span class="co">## try: dataNames, dataYml, dataNotes, dataParams, etc.</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="prepare-public-genomic-data">Prepare public genomic data<a class="anchor" aria-label="anchor" href="#prepare-public-genomic-data"></a>
</h4>
<p>Then we prepare the public genomic data resources that are needed to
run <code>mutect2</code> for somatic variant calling.</p>
<ul>
<li>reference sequence file (<code>b37</code> or <code>hg38</code>)</li>
<li>Panel of Normals (PON) <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890631-Panel-of-Normals-PON-" class="external-link">ref</a>
</li>
</ul>
<p>There are some pre-built data recipes in <code>ReUseData</code> that
are ready to be queried and used. Basically, we run the
<code>recipeUpdate</code> function to synchronize the existing and newly
added recipes (by default the recipes included in <a href="https://github.com/rworkflow/ReUseDataRecipe" class="external-link"><code>ReUseData</code>
package</a>, can also be a user-specific GitHub repository with similar
settings).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## library(ReUseData)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/recipeUpdate.html" class="external-link">recipeUpdate</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Updating recipes...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; recipeHub with 15 records</span></span>
<span><span class="co">#&gt; cache path:  /github/home/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;           name               </span></span>
<span><span class="co">#&gt;   BFC1  | STAR_index         </span></span>
<span><span class="co">#&gt;   BFC2  | bowtie2_index      </span></span>
<span><span class="co">#&gt;   BFC3  | echo_out           </span></span>
<span><span class="co">#&gt;   BFC4  | ensembl_liftover   </span></span>
<span><span class="co">#&gt;   BFC5  | gcp_broad_gatk_hg19</span></span>
<span><span class="co">#&gt;   ...     ...                </span></span>
<span><span class="co">#&gt;   BFC11 | gencode_transcripts</span></span>
<span><span class="co">#&gt;   BFC12 | hisat2_index       </span></span>
<span><span class="co">#&gt;   BFC13 | reference_genome   </span></span>
<span><span class="co">#&gt;   BFC14 | salmon_index       </span></span>
<span><span class="co">#&gt;   BFC15 | ucsc_database</span></span></code></pre></div>
<p>The pre-built recipes of <code>gcp_gatk_mutect2_b37</code> and
<code>gcp_gatk_mutect2_hg38</code> are for the downloading of genomic
data resources from the GATK resource bundle on <a href="https://console.cloud.google.com/storage/browser/gatk-best-practices/somatic-b37" class="external-link">Google
Cloud Bucket</a> based on reference build of <code>b37</code> and
<code>hg38</code> separately. See more details <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890811-Resource-bundle" class="external-link">here</a>.</p>
<p>We will search and load the data recipe, assign values to the
<code>filename</code> and <code>idx</code> parameters, and then evaluate
the recipe (<code>getData</code>) to download the data to your local
machine with automatic tracking and user-specified notes/keywords.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/recipeSearch.html" class="external-link">recipeSearch</a></span><span class="op">(</span><span class="st">"mutect2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; recipeHub with 2 records</span></span>
<span><span class="co">#&gt; cache path:  /github/home/.cache/R/ReUseDataRecipe </span></span>
<span><span class="co">#&gt; # recipeSearch() to query specific recipes using multipe keywords</span></span>
<span><span class="co">#&gt; # recipeUpdate() to update the local recipe cache</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;          name                 </span></span>
<span><span class="co">#&gt;   BFC7 | gcp_gatk_mutect2_b37 </span></span>
<span><span class="co">#&gt;   BFC8 | gcp_gatk_mutect2_hg38</span></span>
<span><span class="va">rcp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/recipeLoad.html" class="external-link">recipeLoad</a></span><span class="op">(</span><span class="st">"gcp_gatk_mutect2_b37"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Note: you need to assign a name for the recipe: rcpName &lt;- recipeLoad('xx')</span></span>
<span><span class="co">#&gt; Data recipe loaded!</span></span>
<span><span class="co">#&gt; Use inputs() to check required input parameters before evaluation.</span></span>
<span><span class="co">#&gt; Check here: https://rcwl.org/dataRecipes/gcp_gatk_mutect2_b37.html</span></span>
<span><span class="co">#&gt; for user instructions (e.g., eligible input values, data source, etc.)</span></span>
<span></span>
<span><span class="co">## get panel of normal vcf</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"Mutect2-exome-panel.vcf"</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="st">"idx"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/getData.html" class="external-link">getData</a></span><span class="op">(</span><span class="va">rcp</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"shareData"</span><span class="op">)</span>,</span>
<span>        notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mutect2"</span>, <span class="st">"panel of normal"</span><span class="op">)</span>,</span>
<span>        showLog <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; }</span></span>
<span><span class="co">#&gt; List of length 4</span></span>
<span><span class="co">#&gt; names(4): command output logs yml</span></span></code></pre></div>
<p>Here Let’s check the output directory for the files.</p>
<ul>
<li>Downloaded genomic file dataset.</li>
<li>Automatically generated meta files.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"shareData"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "gcp_gatk_mutect2_b37_Mutect2-exome-panel.vcf_idx.cwl"</span></span>
<span><span class="co">#&gt; [2] "gcp_gatk_mutect2_b37_Mutect2-exome-panel.vcf_idx.md5"</span></span>
<span><span class="co">#&gt; [3] "gcp_gatk_mutect2_b37_Mutect2-exome-panel.vcf_idx.sh" </span></span>
<span><span class="co">#&gt; [4] "gcp_gatk_mutect2_b37_Mutect2-exome-panel.vcf_idx.yml"</span></span>
<span><span class="co">#&gt; [5] "Mutect2-exome-panel.vcf"                             </span></span>
<span><span class="co">#&gt; [6] "Mutect2-exome-panel.vcf.idx"</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Reference genome files</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"Homo_sapiens_assembly19.fasta"</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">idx</span> <span class="op">=</span> <span class="st">"fai"</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/getData.html" class="external-link">getData</a></span><span class="op">(</span><span class="va">rcp</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"shareData"</span><span class="op">)</span>,</span>
<span>        notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"reference genome"</span>, <span class="st">"hg19"</span>, <span class="st">"b37"</span><span class="op">)</span>,</span>
<span>        showLog <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">filename</span> <span class="op">&lt;-</span> <span class="st">"Homo_sapiens_assembly19.dict"</span></span>
<span><span class="va">rcp</span><span class="op">$</span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/getData.html" class="external-link">getData</a></span><span class="op">(</span><span class="va">rcp</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"shareData"</span><span class="op">)</span>,</span>
<span>        notes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"human"</span>, <span class="st">"reference genome dict"</span>, <span class="st">"hg19"</span>, <span class="st">"b37"</span><span class="op">)</span>,</span>
<span>        showLog <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">## check output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"shareData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we have all the data needed. Here we show some utility functions
to manage, query, and use the data.</p>
<div class="section level5">
<h5 id="data-management">Data management<a class="anchor" aria-label="anchor" href="#data-management"></a>
</h5>
<p><code>dataUpdate</code> will manage the data downloaded so that the
data can be managed and easily searched by <code>dataSearch</code> with
other convenient utility fuctions.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataUpdate.html" class="external-link">dataUpdate</a></span><span class="op">(</span>dir <span class="op">=</span> <span class="va">workdir</span>, keepTags<span class="op">=</span><span class="cn">FALSE</span>, cleanup <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">rs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataSearch.html" class="external-link">dataSearch</a></span><span class="op">(</span><span class="st">"mutect2"</span><span class="op">)</span></span>
<span><span class="va">rs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataSearch.html" class="external-link">dataSearch</a></span><span class="op">(</span><span class="st">"reference"</span><span class="op">)</span></span></code></pre></div>
<p>Data can be easily tagged e.g., with a specific software name, so
that it can be retrieved easily after with that keyword using
<code>dataSearch</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataTags</a></span><span class="op">(</span><span class="va">rs2</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"mutect2"</span></span>
<span><span class="va">rs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataSearch.html" class="external-link">dataSearch</a></span><span class="op">(</span><span class="st">"mutect2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataNames</a></span><span class="op">(</span><span class="va">rs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataNotes</a></span><span class="op">(</span><span class="va">rs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataParams</a></span><span class="op">(</span><span class="va">rs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataTags</a></span><span class="op">(</span><span class="va">rs</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataPaths</a></span><span class="op">(</span><span class="va">rs</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">pon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ReUseData/man/dataHub-class.html" class="external-link">dataPaths</a></span><span class="op">(</span><span class="va">rs</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="run-mutect2-pipeline-in-r">Run <code>mutect2</code> pipeline in R<a class="anchor" aria-label="anchor" href="#run-mutect2-pipeline-in-r"></a>
</h4>
<div class="section level5">
<h5 id="load-tools-or-pipeline">Load tools or pipeline<a class="anchor" aria-label="anchor" href="#load-tools-or-pipeline"></a>
</h5>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RcwlPipelines</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlUpdate.html" class="external-link">cwlUpdate</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlSearch.html" class="external-link">cwlSearch</a></span><span class="op">(</span><span class="st">"mutect2"</span><span class="op">)</span></span>
<span><span class="va">mutect2pl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html" class="external-link">cwlLoad</a></span><span class="op">(</span><span class="st">"pl_Mutect2PL"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/plotCWL.html" class="external-link">plotCWL</a></span><span class="op">(</span><span class="va">mutect2pl</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## run simple tool instead of pipeline because of large files</span></span>
<span><span class="va">mutect2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/cwlLoad.html" class="external-link">cwlLoad</a></span><span class="op">(</span><span class="st">"tl_Mutect2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## change to smaller docker image</span></span>
<span><span class="va">ds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/RcwlPipelines/man/searchContainer.html" class="external-link">searchContainer</a></span><span class="op">(</span><span class="st">"gatk4"</span><span class="op">)</span></span>
<span><span class="va">req1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwl-requirements.html" class="external-link">requireDocker</a></span><span class="op">(</span><span class="va">ds</span><span class="op">[</span><span class="fl">1</span>, <span class="st">"container"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">requirements</a></span><span class="op">(</span><span class="va">mutect2</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">req1</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## check inputs</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/cwlProcess-methods.html" class="external-link">inputs</a></span><span class="op">(</span><span class="va">mutect2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## prepare capture region</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">21</span>, <span class="fl">10400000</span>, <span class="fl">10500000</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"region.bed"</span><span class="op">)</span>,</span>
<span>            row.names<span class="op">=</span><span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, sep<span class="op">=</span><span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## assign inputs</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">tbam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"data/tumor.bam"</span><span class="op">)</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">nbam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"data/normal.bam"</span><span class="op">)</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">normal</span> <span class="op">&lt;-</span> <span class="st">"NA12892"</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">Ref</span> <span class="op">&lt;-</span> <span class="va">ref</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">pon</span> <span class="op">&lt;-</span> <span class="va">pon</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">interval</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"region.bed"</span><span class="op">)</span></span>
<span><span class="va">mutect2</span><span class="op">$</span><span class="va">out</span> <span class="op">&lt;-</span> <span class="st">"somatic.vcf"</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## run tool</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/Rcwl/man/runCWL.html" class="external-link">runCWL</a></span><span class="op">(</span><span class="va">mutect2</span>, outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"output"</span><span class="op">)</span>, docker <span class="op">=</span> <span class="st">"udocker"</span>, showLog <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## checkout pipeline results</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">workdir</span>, <span class="st">"output"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Roswell Park Comprehensive Cancer Center<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Roswell Park Comprehensive Cancer Center<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Qian Liu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
